File: 965.Univalued-Binary-Tree.py
# https://leetcode.com/problems/univalued-binary-tree/
# Easy (73.15%)
# Total Accepted:    6,213
# Total Submissions: 8,494
# beats 91.04% of python submissions

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def __init__(self):
        self.value = None
        
    def isUnivalTree(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True
        
        if self.value is None:
            self.value = root.val
        
        elif root.val != self.value:
            return False
        
        return self.isUnivalTree(root.left) and self.isUnivalTree(root.right)
    

File: 965.Univalued-Binary-Tree.java
// https://leetcode.com/problems/univalued-binary-tree/
//
// algorithms
// Easy (66.96%)
// Total Accepted: 60,518
// Total Submissions: 90,384
// beats 100.0% of java submissions


/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode(int x) { val = x; } }
 */
class Solution {

    private static boolean res;

    public boolean isUnivalTree(TreeNode root) {
        if (root == null) {
            return true;
        }

        res = true;
        recursive(root, root.val);

        return res;
    }

    public void recursive(TreeNode root, int value) {
        if (root == null || !res) {
            return;
        }

        if (root.val != value) {
            res = false;
            return;
        }

        recursive(root.left, value);
        recursive(root.right, value);
    }

}